presentation/src/main/java/com/github/ianellis/shifts/presentation/ShiftsViewModel.kt

```diff
@@ -36,42 +40,42 @@ class ShiftsViewModel constructor(
     private var shiftUpdateJob: Job? = null
 
     init {
-        loadShifts()
-    }
+        shifts.value = emptyList()
+        shiftState.value = CurrentShiftState.OFF
 
-    private val onShiftsChanged: (Event<List<Shift>>) -> Unit = { event ->
         launch {
-            event.value?.let {
-                shifts.value = it
-                if (shifts.value?.lastOrNull()?.state is ShiftState.Started) {
-                    shiftState.value = CurrentShiftState.STARTED
+            loadShifts()
+        }
+
+        launch {
+            shiftsUpdated.consumeEach { event ->
+
+                event.value?.let {
+                    shifts.value = it
+                    if (shifts.value?.lastOrNull()?.state is ShiftState.Started) {
+                        shiftState.value = CurrentShiftState.STARTED
+                    }
                 }
-            }
 
-            event.error?.let {
-                showGenericError.value = event.error
-            }
+                event.error?.let {
+                    showGenericError.value = event.error
+                }
 
-            loading.value = false
+                loading.value = false
+            }
         }
-    }
 
-    private val ignoreShifts: Ignore
-
-    init {
-        shifts.value = emptyList()
-        shiftState.value = CurrentShiftState.OFF
-        ignoreShifts = listenForShifts(onShiftsChanged)
         loading.value = true
     }
 
     fun startShift(v: View? = null) {
         if (shiftUpdateJob?.isActive != true && shiftState.value == CurrentShiftState.OFF) {
```

Comment:
I am not sure why we need to make these extra checks here. From UI point of view, it should not be possible to execute `startShift` twice or execute `endShift` while `startShift` is in progress.

If it's not handled on UI side, those operations should be put into some kind of queue. With the current approach, the second operation will be just ignored silently while the first is in progress.

Reply:
I want the second operation to be ignored. You are right that from the UI point of view this should not happen - but this VM should not be coupled to the views implementation. This extra check protects us from someone mis-implementing the view

Reply:
> This extra check protects us from someone mis-implementing the view

I think this would just mask the issue. If someone mis-implement the view they will never know that 🤔 

So, I think if we really want to make ViewModel safe from misuse, it's better to make a lock which will put all the operations in queue.

E.g.
```kotlin
val mutext = Mutex()

fun startShift() {
    mutext.withLock { 
        if (shiftState.value == CurrentShiftState.OFF) { // rest of the code }
    }
}

fun endShift() {
    mutext.withLock { 
        if (shiftState.value == CurrentShiftState.STARTED) { // rest of the code }
    }
}
```